src/pdf.sandbox.external.js

```diff
@@ -0,0 +1,177 @@
+/* Copyright 2020 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// In mozilla-central, this file is loaded as non-module script,
+// so it mustn't have any dependencies.
+
+class SandboxSupportBase {
+  /**
+   * @param {DOMWindow} - win
+   */
+  constructor(win) {
+    this.win = win;
+    this.timeoutIds = new Map();
+
+    // Will be assigned after the sandbox is initialized
+    this.commFun = null;
+  }
+
+  destroy() {
+    this.commFunc = null;
+    this.timeoutIds.forEach(([_, id]) => this.win.clearTimeout(id));
+    this.timeoutIds = null;
+  }
+
+  /**
+   * @param {Object} val - Export a value in the sandbox.
+   */
+  exportValueToSandbox(val) {
+    throw new Error("Not implemented");
+  }
+
+  /**
+   * @param {Object} val - Import a value from the sandbox.
+   */
+  importValueFromSandbox(val) {
+    throw new Error("Not implemented");
+  }
+
+  /**
+   * @param {Object} err - Report an error.
+   */
+  reportError(err) {
+    throw new Error("Not implemented");
+  }
+
+  /**
+   * @param {Object} val - Import a value from the sandbox.
+   */
+  callSandboxFunction(name, args) {
+    try {
+      args = this.exportValueToSandbox(args);
+      this.commFun(name, args);
+    } catch (e) {
+      this.reportError(e);
```

Comment:
Doesn't just `console.error(e);` work for both m-c and the web app? Let's do that instead. Keeping this file minimal makes it easier to audit.